File: 2024/290. Word Pattern.cpp
class Solution {
public:
    bool wordPattern(string pattern, string s) {
        unordered_map<char, int> pm;
        unordered_map<string, int> sm;
        int pos = 0, prev = -1, index = 0;
        while(pos < s.length() && index < pattern.length()){
            pos = s.find(" ", prev + 1);
            string word = s.substr(prev + 1, pos - prev - 1);
            if (pm.find(pattern[index]) == pm.end()) {
                pm[pattern[index]] = index;
            }
            if (sm.find(word) == sm.end()) {
                sm[word] = index;
            }
            if (pm[pattern[index]] != sm[word]) {
                return false;
            }
            index += 1;
            prev = pos;
        }
        return index >= pattern.length() && pos >= s.length();
    }
};

File: 2024/290. Word Pattern.py
class Solution:
    def wordPattern(self, pattern: str, s: str) -> bool:
        words = s.split(' ')
        if len(words) != len(pattern):
            return False
        p_map = defaultdict(int)
        s_map = defaultdict(int)
        for i in range(len(pattern)):
            if p_map[pattern[i]] != s_map[words[i]]:
                return False
            p_map[pattern[i]] = i + 1
            s_map[words[i]] = i + 1
        return True
